cabal-install/Distribution/Client/IndexUtils.hs

```diff
@@ -307,12 +307,15 @@ extractPkg entry blockNo = case Tar.entryContent entry of
   Tar.OtherEntryType typeCode content _
     | Tar.isBuildTreeRefTypeCode typeCode ->
       Just $ do
-        let path   = byteStringToFilePath content
-            err = "Error reading package index."
-        cabalFile <- tryFindAddSourcePackageDesc path err
-        descr     <- PackageDesc.Parse.readPackageDescription normal cabalFile
-        return $ BuildTreeRef (refTypeFromTypeCode typeCode) (packageId descr)
-                              descr path blockNo
+        let path = byteStringToFilePath content
```

Comment:
So the problem is that `extractPkg` is called from `parsePackageIndex`, which is called from `updatePackageIndexCacheFile`, which is called from `removeBuildTreeRefs`. But isn't `updatePackageIndexCacheFile` called only after `doRemove` - that is, after the invalid refs were already removed? So why should we fix `extractPkg` then - maybe it just suffices to improve the error message?